subsys/testsuite/ztest/include/ztest_test_new.h

```diff
@@ -0,0 +1,323 @@
+ */
+int z_ztest_run_test_suite(const char *name);
+
+/**
+ * @defgroup ztest_test Ztest testing macros
+ * @ingroup ztest
+ *
+ * This module eases the testing process by providing helpful macros and other
+ * testing structures.
+ *
+ * @{
+ */
+
+/**
+ * @brief Fail the currently running test.
+ *
+ * This is the function called from failed assertions and the like. You
+ * probably don't need to call it yourself.
+ */
+void ztest_test_fail(void);
+
+/**
+ * @brief Pass the currently running test.
+ *
+ * Normally a test passes just by returning without an assertion failure.
+ * However, if the success case for your test involves a fatal fault,
+ * you can call this function from k_sys_fatal_error_handler to indicate that
+ * the test passed before aborting the thread.
+ */
+void ztest_test_pass(void);
+
+/**
+ * @brief Skip the current test.
+ *
+ */
+void ztest_test_skip(void);
+
+/**
+ * @brief Do nothing, successfully.
+ *
+ * Unit test / setup function / teardown function that does
+ * nothing, successfully. Can be used as a parameter to
+ * ztest_unit_test_setup_teardown().
+ */
+static inline void unit_test_noop(void)
+{
+}
+
+#define Z_ZTEST(suite, fn, t_options)                                                              \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(void);                                                            \
+	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
+		.test_suite_name = STRINGIFY(suite),                                               \
+		.name = STRINGIFY(fn),                                                             \
+		.test = (_##suite##_##fn##_wrapper),                                               \
+		.thread_options = t_options,                                                       \
+	};                                                                                         \
+	static void _##suite##_##fn##_wrapper(void *data)                                          \
+	{                                                                                          \
+		ARG_UNUSED(data);                                                                  \
+		suite##_##fn();                                                                    \
+	}                                                                                          \
+	static inline void suite##_##fn(void)
+
+#define Z_ZTEST_F(suite, fn, t_options)                                                            \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(struct suite##_fixture *this);                                    \
+	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
+		.test_suite_name = STRINGIFY(suite),                                               \
```

Comment:
this seems redundant as suite has name.

Reply:
The suite name is how we match the unit tests to the suite, in what way is it redundant? Is there an optimization I missed?

Reply:
I didn't look into the implementation but thought that section is used to identify test cases.

Reply:
So the way this is currently implemented is:

* There's  a section for suites that contains:
  * Name string
  * An optional predicate function to conditionally run the suite
  * A setup function to run before the suite; optionally returning a fixture
  * A before function to run before each unit test
  * An after function to run after each unit test
  * A teardown function to run after all the unit tests
* A section for unit tests. Unit tests contain the suite's name as a string. It's possible that we could optimize this a bit later by augmenting the post-build script, but for now it's just using a simple `strcmp` to figure out which unit tests to run for which suite. What this buys us is the ability for the suite to be declared in one compilation unit (`.c` file) and the unit tests in another (or even multiple). This is especially helpful with twister as a core set of unit tests can be in one `.c` file and the rest may be conditionally compiled based on `CONFIG_` values.